src/zeroD/ReactorNet.cpp

```diff
@@ -471,22 +471,59 @@ void ReactorNet::preconditionerSetup(double t, double* y, double gamma)
     precon->setup();
 }
 
-void ReactorNet::checkPreconditionerSupported()
+void ReactorNet::updatePreconditioner(double gamma)
 {
-    // preconditioner currently not supported for surfaces
-    for (size_t i = 0; i < m_reactors.size(); i++) {
-        if (m_reactors[i]->nSurfs() > 0) {
-            throw CanteraError("ReactorNet::checkPreconditionerSupported",
-                "Preconditioning is not supported for networks with surfaces.");
+    auto precon = m_integ->preconditioner();
+    precon->setGamma(gamma);
+    precon->updatePreconditioner();
+}
+
+Eigen::SparseMatrix<double> ReactorNet::jacobian()
+{
+    // network must be initialized for the jacobian
+    if (! m_init) {
+        initialize();
+    }
+    // create system jacobian from reactor jacobians
+    std::vector<Eigen::Triplet<double>> jac_trips;
+    Eigen::SparseMatrix<double> system_jac(m_nv, m_nv);
+    for (size_t n = 0; n < m_reactors.size(); n++) {
+        Reactor& r = *m_reactors[n];
+        Eigen::SparseMatrix<double> r_jac = r.jacobian();
+        size_t st = m_start[n];
+        for (int k=0; k<r_jac.outerSize(); ++k) {
+            for (Eigen::SparseMatrix<double>::InnerIterator it(r_jac, k); it; ++it) {
+                jac_trips.emplace_back(it.row() + st, it.col() + st, it.value());
+            }
         }
     }
+    // set jacobian from triples
+    system_jac.setFromTriplets(jac_trips.begin(), jac_trips.end());
+    return system_jac;
 }
 
-void ReactorNet::updatePreconditioner(double gamma)
+Eigen::SparseMatrix<double> ReactorNet::finiteDifferenceJacobian()
 {
-    auto precon = m_integ->preconditioner();
-    precon->setGamma(gamma);
-    precon->updatePreconditioner();
+    // network must be initialized for the jacobian
+    if (! m_init) {
+        initialize();
+    }
+    // create system jacobian from reactor jacobians
+    std::vector<Eigen::Triplet<double>> jac_trips;
+    Eigen::SparseMatrix<double> system_fd_jac(m_nv, m_nv);
+    for (size_t n = 0; n < m_reactors.size(); n++) {
+        Reactor& r = *m_reactors[n];
+        Eigen::SparseMatrix<double> r_fd_jac = r.finiteDifferenceJacobian();
+        size_t st = m_start[n];
+        for (int k=0; k<r_fd_jac.outerSize(); ++k) {
+            for (Eigen::SparseMatrix<double>::InnerIterator it(r_fd_jac, k); it; ++it) {
+                jac_trips.emplace_back(it.row() + st, it.col() + st, it.value());
```

Comment:
I will reconsider my math for this but in my original derivation I treated inflow/outlfow as a constant source term in the species equations which would just go to zero when you differentiate. I suppose the inflow doesn't have to be constant though but would it ever be a function of other species? Not quite sure how to approach this. I will consider it though.

Reply:
I think inflow / outflow from/to reservoirs is mostly fine. But if you have reactors connected to one another, then $dn_k/dt$ of a species in the downstream reactor is a function of $n_k$ in the upstream reactor, assuming the mass flow rate is constant. 

Strictly speaking, I think there should probably also be something with total mass flow rate, which depends on the pressure difference between the reactors (at least according to the model of the `Valve` object), and therefore dependent on all of the $n_k$ in both reactors, though this may be one of the terms that you want to partially neglect in the preconditioner.